### ISSUE TITLE
Support GHC 8.4.1

### ISSUE THREAD

**Issue by asr** [agda/agda#2878]
Last edit:  2018-03-28

GHC 8.4.1 was [released](https://mail.haskell.org/pipermail/ghc-devs/2018-March/015452.html). I created this issue for collecting/reporting the required changes/issues.

Some fixed issues (for the record):

* cabal-install

  `make install-bin` generates the following error (I filed [this](https://ghc.haskell.org/trac/ghc/ticket/14775) GHC issue):

  ```
  Failed to install Agda-2.5.4
  cabal: Error: some packages failed to install:
  Agda-2.5.4-KXHLQFOYJSo9BesMhydjzQ failed during the final install step. The
  exception was:
  dieVerbatim: user error (cabal: Couldn't parse the output of 'setup register
  --gen-pkg-config':NoParse
  "license" 5
  )
  ```

  The issue was fixed by `cabal-install` 2.2.0.0.

* agda (https://github.com/agda/agda/issues/2957)
* async (https://github.com/simonmar/async/issues/77)
* blaze-builder (https://github.com/lpsmith/blaze-builder/pull/10)
* blaze-markup (https://github.com/jaspervdj/blaze-markup/issues/33)
* boxes (https://github.com/treeowl/boxes/issues/29 and https://github.com/treeowl/boxes/pull/30)
* EdisonCore (https://github.com/robdockins/edison/issues/11 and https://github.com/robdockins/edison/pull/12)
* exceptions (https://github.com/ekmett/exceptions/pull/61)
* fmlist (https://github.com/sjoerdvisscher/fmlist/issues/2)
* generic-deriving (https://github.com/dreixel/generic-deriving/issues/51)
* geniplate (https://github.com/danr/geniplate/pull/9)
* happy (https://github.com/simonmar/happy/issues/113)
* hashable (https://github.com/tibbe/hashable/pull/150)
* hashtables (the issue https://github.com/gregorycollins/hashtables/issues/45 and the pull request https://github.com/gregorycollins/hashtables/pull/46)
* listlike (https://github.com/JohnLato/listlike/pull/12)
* pretty-show (required by `shelltestrunner`) (https://github.com/yav/pretty-show/pull/25)
* primitive (https://github.com/haskell/primitive/issues/72)
* process-extras (https://github.com/seereason/process-extras/pull/13)
* regex-tdfa (https://github.com/ChrisKuklewicz/regex-tdfa/pull/16)
* shelltestrunner (https://github.com/simonmichael/shelltestrunner/issues/5)
* split (https://github.com/byorgey/split/pull/3)
* tagged (https://github.com/ekmett/tagged/pull/42)
* tasty (https://github.com/feuerbach/tasty/issues/196)
* tasty-silver (https://github.com/phile314/tasty-silver/issues/14)
* unordered-containers (https://github.com/tibbe/unordered-containers/issues/174)
* vector (https://github.com/haskell/vector/pull/197)

**Comment by asr** [comment]
GHC 8.4.1 was [released](https://mail.haskell.org/pipermail/ghc-devs/2018-March/015452.html). The OP shows the current status for supporting this version of GHC.

**Comment by asr** [comment]
Could someone test on Windows the library Win32 2.6.1.0 [shipped](https://downloads.haskell.org/~ghc/8.4.1/docs/html/users_guide/8.4.1-notes.html#included-libraries) with GHC 8.4.1, please. Note that this version of Win32 is [ backwards incompatible](https://downloads.haskell.org/~ghc/8.4.1/docs/html/users_guide/8.4.1-notes.html#win32).

Note also that due to the `cabal-install` issue reported in the OP, at the moment it's necessary comment-out the `license` field in `Agda.cabal`.

**Comment by asr** [comment]
FYI.

After running `cabal update` (because a new version of hashtables was released),  `stable-2.5` and `master` compile with GHC 8.4.1 (tested on Ubuntu 16.04).

TODO: See the OP.

**Comment by asr** [comment]
Could someone test GHC 8.4.1 on Windows, please.

Note that the version of the  `Win32` library shipped with GHC 8.4.1 is [ backwards incompatible](https://downloads.haskell.org/~ghc/8.4.1/docs/html/users_guide/8.4.1-notes.html#win32).

**Comment by asr** [comment]
Note that GHC 8.4.1 requires `cabal-install` ≥ 2.2.0.0.

It seems you can install `cabal-install` 2.2.0.0 using at least the following options:

1. On GHC 8.4.1: Bootstrap it.

2. On  GHC <= 8.2.2: Use the previous version installed of `cabal-install`.

(I used the second option).
